go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+type blameGraph map[hash.Hash]blameInfo
+
+// Blame implements the `dolt blame` command. Blame annotates each row in the given table with information
+// from the revision which last modified the row, optionally starting from a given revision.
+//
+// Blame is computed as follows:
+//
+// First, a blame graph is initialized with one node for every row in the table at the given commit (defaulting
+// to HEAD of the currently checked-out branch).
+//
+// Starting from the given commit, walk backwards through the commit graph (currently by following each commit's
+// first parent, though this may change in the future).
+//
+// For each adjacent pair of commits `old` and `new`, check each remaining unblamed node to see if the row it represents
+// changed between the commits. If so, mark it with `new` as the blame origin and continue to the next node without blame.
+//
+// When all nodes have blame information, stop iterating through commits and print the blame graph.
+func Blame(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := argparser.NewArgParser()
+	help, usage := cli.HelpAndUsagePrinters(commandStr, blameShortDesc, blameLongDesc, blameSynopsis, ap)
+	apr := cli.ParseArgs(ap, args, help)
+
+	if apr.NArg() < 2 {
+		usage()
+		return 1
+	}
+
+	cs, err := parseCommitSpec(dEnv, apr)
+	if err != nil {
+		cli.PrintErr(err)
+		return 1
+	}
+
+	tableName := apr.Arg(1)
+
+	if err := runBlame(ctx, dEnv, cs, tableName); err != nil {
+		cli.PrintErr(err)
+		return 1
+	}
+
+	return 0
+}
+
+func runBlame(ctx context.Context, dEnv *env.DoltEnv, cs *doltdb.CommitSpec, tableName string) error {
+	commit, err := dEnv.DoltDB.Resolve(ctx, cs)
+	if err != nil {
+		return err
+	}
+
+	blameGraph, err := blameGraphFromCommit(ctx, dEnv, commit, tableName)
+	if err != nil {
+		return err
+	}
+
+	pkColNames, err := pkColNamesFromCommit(ctx, commit, tableName)
+	if err != nil {
+		return err
+	}
+
+	cli.Println(blameGraph.String(ctx, pkColNames))
+	return nil
+}
+
+func blameGraphFromCommit(ctx context.Context, dEnv *env.DoltEnv, commit *doltdb.Commit, tableName string) (*blameGraph, error) {
+	// get the commits sorted from newest to oldest ending with `commit`
+	commits, err := actions.TimeSortedCommits(ctx, dEnv.DoltDB, commit, -1)
+	if err != nil {
+		return nil, err
+	}
+
+	rows, err := rowsFromCommit(ctx, commit, tableName)
+	if err != nil {
+		return nil, err
+	}
+
+	blameGraph, err := blameGraphFromRows(ctx, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range commits {
+		// get the first parent
+		parent, err := dEnv.DoltDB.ResolveParent(ctx, c, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		// for each unblamed node, see if it changed between `c` and `parent`.
+		// if so, mark it blamed with `c` as the blame origin
+		for _, rowPK := range blameGraph.UnblamedNodeKeys() {
+			changed, err := rowChanged(ctx, parent, c, tableName, rowPK)
```

Comment:
Should we be comparing the value of the row in `HEAD` against the value of the row in this commit, and taking the last commit where the value is the same? Is the most recent time sorted commit where the value of this node changed against its parent the same thing? Maybe they only ever differ for intermediate schema changes where the columns that are in `c` and `parent` aren't in `HEAD` anymore...?

Reply:
I think it should be equivalent. If the blame of a row is the most recent commit where the row changed against its parent, and schema changes count as changes to every row (I am not confident that this is actually what happens but I think it's what we want, right?), then intermediate schema changes like this are precluded.

Am I missing something?